File: iOS/ProsaryTests/MultiDayRunTests.swift
//
//  MultiDayRunTests.swift
//  ProsaryTests
//
//  The rules that make a multi-day devotion behave like a calendar rather than a counter.
//

import XCTest
@testable import Prosary

final class MultiDayRunTests: XCTestCase {
  private let nine = 9
  private func day(_ offset: Int, from start: Date) -> Date {
    Calendar.current.date(byAdding: .day, value: offset, to: start)!
  }

  func testPrayingTwiceInOneDayDoesNotAdvance() {
    let start = Date()
    var run = MultiDayRun(devotionId: "novena", startedOn: start)
    run.recordPrayed(day: 0, on: start)

    XCTAssertTrue(run.hasPrayedToday(on: start))
    // Same calendar day: the next unprayed day exists, but today's is already done.
    XCTAssertEqual(run.dueDay(dayCount: nine, on: start), 0)
    XCTAssertEqual(run.nextUnprayedDay(dayCount: nine), 1)
  }

  func testTheNextCalendarDayOffersTheNextDay() {
    let start = Date()
    var run = MultiDayRun(devotionId: "novena", startedOn: start)
    run.recordPrayed(day: 0, on: start)

    let tomorrow = day(1, from: start)
    XCTAssertFalse(run.hasPrayedToday(on: tomorrow))
    XCTAssertEqual(run.resumption(dayCount: nine, on: tomorrow), .resume(day: 1))
    XCTAssertNil(run.missedDay(dayCount: nine, on: tomorrow))
  }

  func testAMissedDayOffersBothItAndTheCalendarDay() {
    let start = Date()
    var run = MultiDayRun(devotionId: "novena", startedOn: start)
    run.recordPrayed(day: 0, on: start)

    // Day 2 (index 1) was due yesterday and never prayed; today the calendar wants index 2.
    let dayAfterNext = day(2, from: start)
    XCTAssertEqual(run.missedDay(dayCount: nine, on: dayAfterNext), 1)
    XCTAssertEqual(run.dueDay(dayCount: nine, on: dayAfterNext), 2)
    XCTAssertEqual(run.resumption(dayCount: nine, on: dayAfterNext), .choose(missed: 1, next: 2))
  }

  func testTheDayCountComesFromTheDevotion() {
    let start = Date()
    var triduum = MultiDayRun(devotionId: "triduum", startedOn: start)
    triduum.recordPrayed(day: 0, on: start)
    triduum.recordPrayed(day: 1, on: day(1, from: start))
    XCTAssertFalse(triduum.isComplete(dayCount: 3))

    triduum.recordPrayed(day: 2, on: day(2, from: start))
    XCTAssertTrue(triduum.isComplete(dayCount: 3))
    XCTAssertEqual(triduum.resumption(dayCount: 3, on: day(2, from: start)), .complete)
    // The same run is nowhere near done if the devotion is longer.
    XCTAssertFalse(triduum.isComplete(dayCount: 33))
  }

  func testDueDayNeverRunsPastTheLastDay() {
    let start = Date()
    let run = MultiDayRun(devotionId: "novena", startedOn: start)
    XCTAssertEqual(run.dueDay(dayCount: nine, on: day(400, from: start)), 8)
  }
}

File: Windows/Prosary.Tests/MultiDayRunTests.cs
using Prosary.Models;
using Xunit;

namespace Prosary.Tests;

/// The rules that make a multi-day devotion behave like a calendar rather than a counter.
public class MultiDayRunTests
{
    private const int Nine = 9;
    private static readonly DateTimeOffset Start = new(2026, 8, 5, 9, 0, 0, TimeSpan.Zero);
    private static DateTimeOffset Day(int offset) => Start.AddDays(offset);

    [Fact]
    public void PrayingTwiceInOneDayDoesNotAdvance()
    {
        var run = new MultiDayRun("novena", Start, []).RecordingPrayed(0, Start);
        Assert.True(run.HasPrayedToday(Start));
        Assert.Equal(0, run.DueDay(Nine, Start));
        Assert.Equal(1, run.NextUnprayedDay(Nine));
    }

    [Fact]
    public void TheNextCalendarDayOffersTheNextDay()
    {
        var run = new MultiDayRun("novena", Start, []).RecordingPrayed(0, Start);
        Assert.False(run.HasPrayedToday(Day(1)));
        Assert.Equal(new MultiDayRun.Resumption.Resume(1), run.GetResumption(Nine, Day(1)));
        Assert.Null(run.MissedDay(Nine, Day(1)));
    }

    [Fact]
    public void AMissedDayOffersBothItAndTheCalendarDay()
    {
        var run = new MultiDayRun("novena", Start, []).RecordingPrayed(0, Start);
        Assert.Equal(1, run.MissedDay(Nine, Day(2)));
        Assert.Equal(2, run.DueDay(Nine, Day(2)));
        Assert.Equal(new MultiDayRun.Resumption.Choose(1, 2), run.GetResumption(Nine, Day(2)));
    }

    [Fact]
    public void TheDayCountComesFromTheDevotion()
    {
        var run = new MultiDayRun("triduum", Start, [])
            .RecordingPrayed(0, Start).RecordingPrayed(1, Day(1)).RecordingPrayed(2, Day(2));
        Assert.True(run.IsComplete(3));
        Assert.Equal(new MultiDayRun.Resumption.Complete(), run.GetResumption(3, Day(2)));
        Assert.False(run.IsComplete(33));
    }

    [Fact]
    public void DueDayNeverRunsPastTheLastDay() =>
        Assert.Equal(8, new MultiDayRun("novena", Start, []).DueDay(Nine, Day(400)));
}
